feat: replace open in query builder with new insight page

pauldambra opened the pull request:
## Problem

Now that the insight scene can be used to edit insights, let's open query editing in a new insight 

see #13927

## Changes

does that

![2023-03-13 14 27 54](https://user-images.githubusercontent.com/984817/224739069-bacd92c8-29a4-4357-89b6-8a355db36fa1.gif)

After https://github.com/PostHog/posthog/pull/14688 we can teach the insight scene to open the correct tab depending on if a hogql or non-hogql query is being presented

Will follow-up with switching insight viz tabs to text editing...

## How did you test this code?

👀 locally

pauldambra committed:
feat: replace open in query builder with new insight page

thmsobrmlr commented:
Not sure what's in scope for this PR yet, but I'd expect that clicking the button selects the query tab and closes the drawer. Code wise all looks good 👍 

Not for now, just wanted to mention: It would be great to pass the query not by url, but through some router state, which I think isn't possible yet. I suspect we'll run into URL length limit with this otherwise.

mariusandra commented:
I was thinking we should get rid of the drawer altogether. In my imagination the "<>" button on "event explorer" directly opens a new insight. 🤔

Then, I'd hide that button when editing an insight itself. For some views, we could have the "<>" edit query drawer open, but I'd hide that under "..." -> "edit source".

thmsobrmlr commented:
I'm not a fan of the <> either, so +1 for getting rid of that here. 

Opens up the question of how we'd edit the query for an insight then though. In a previous design decision we opted for making the tabs in insights independent of each other, so that with data exploration each tab will open in a "blank" state. It would be strange if the query tab would behave differently.

I could imagine a small toggle in the top-right corner of editor filters that switches from filters to a query view and vice-versa.

pauldambra committed:
no drawer

pauldambra commented:
> Opens up the question of how we'd edit the query for an insight then

Yep, and at the moment we check "is this an insight viz" and if so you can't get to the query tab! So, we'd deffo need a different route. 

I like the idea of swapping between text-based editing and click-based editing 🤔

pauldambra committed:
no edit source button at top of insights

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

mariusandra commented:
Summarising this, what about a flow where:

1. Everywhere we have any embedded "query", be it the events explorer, the persons table, a groups table, a random embedded insight in an experiment, a persons modal, or whatever... then instead of `<>`, it'll have a `:open_in_new_window:` icon, which will open it as a new insight.
2. When in the "insights" view, the same icon will be replaced with a `<>` "edit source" button, which for now will just open the drawer to edit the JSON. For now. Perhaps we'll replace it with a non-drawer later.
3. The insights view has tabs, one for each "insight type". We'll add two more: "SQL" and "Data table". One for HogQL queries (DataTableNode -> HogQLQuery), one for any other data table. When you open anything "as a new insight", we'll directly open the right tab. Be it "Trends", "SQL" or "Data table".
4. Any unknown query types can create a new tab... or just open the data table.
5. In the future we'll convert data tables to "[universal search](https://github.com/PostHog/posthog/issues/7963)" tables where you can also select the data source (events? persons? anything!)... but we don't care about that today...

pauldambra committed:
draw the rest of the owl

pauldambra committed:
no need to set setQuery

pauldambra committed:
a little saferety

pauldambra committed:
a little testing

pauldambra committed:
refactor to make it clearer what is changing between tests

pauldambra committed:
remove commented code

pauldambra committed:
include default in schema

pauldambra commented:
updated the gif in the description

![2023-03-13 14 27 54](https://user-images.githubusercontent.com/984817/224739161-1603a112-c825-49be-8160-32159c134413.gif)

pauldambra commented:
Didn't see your comment @mariusandra, so I have a flow

1. Everywhere we have any embedded "query" *that is not an insight viz* then instead of <>, it'll have a :edit database: icon, which will open it as a new insight.
2. When in that "insights" view, we are editing as text so it opens in "query" tab (as we add more tabs we can direct people to the correct one instead e.g. "SQL" for hogQL.

I think that at the moment we need the fall back to "this is a query" and then we can start to open the correct tab as we solidify this

The line between an `InsightViz` and a `Query` is pretty strict at the moment. 

Hmmm, but I did remove the drawer entirely here which gives you no escape hatch for insights 🤔

pauldambra committed:
push query editor down into query

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
allow showing an editor panel for insight viz queries

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
only show query editor button when using data exploration

pauldambra committed:
schema build

mariusandra reviewed (commented):
1. Can this get the non-web-standard tooltip that would say "Open as a new insight"?
<img width="177" alt="image" src="https://user-images.githubusercontent.com/53387/224944147-de92afb1-a4c7-4e15-a181-ad8c47ab3f15.png">

2. Something's still flakey when clicking this in the event explorer:
![2023-03-14 09 41 54](https://user-images.githubusercontent.com/53387/224944568-7a5ce829-fdc9-44ca-b016-ae3a624bc47b.gif)

mariusandra commented:
Totally worth handling later, but instead of "QueryKind: DataTableNode", this autogenerated title could say something like "SQL Query" or "data table events" or something cleaner :).

<img width="747" alt="image" src="https://user-images.githubusercontent.com/53387/224945137-6ad4ab7e-04d9-461a-9614-cf5642886039.png">

pauldambra committed:
fancy tooltip

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
make the tab changing and loading work more betterer

pauldambra commented on frontend/src/scenes/insights/insightDataLogic.ts:
```diff
@@ -250,19 +256,20 @@ export const insightDataLogic = kea<insightDataLogicType>([
-                },
-                {}
-            )
+            // todo what about results that aren't in this shape?
```
@thmsobrmlr to check I'm not missing things here ... 

spreading `values.insight` and calling `setInsight` here was collapsing the inflightFilters and things.

I think all we need to do is update the results we just received back but....

pauldambra commented on frontend/src/scenes/insights/insightDataLogic.ts:
```diff
@@ -250,19 +256,20 @@ export const insightDataLogic = kea<insightDataLogicType>([
-                },
-                {}
-            )
+            // todo what about results that aren't in this shape?
+            actions.setInsightResults({
+                result: insightData?.result,
+                next: insightData?.next,
+            })
```
@thmsobrmlr to check I'm not missing things here ...

spreading values.insight and calling setInsight here was collapsing the inflightFilters and things.

I think all we need to do is update the results we just received back but....

thmsobrmlr replied:
At the moment we treat some of the charts as black boxes in the sense that the internals aren't converted yet and still rely on legacy filters. Not setting filters will break these charts unfortunately. An example where this seems to be the case is the retention chart that breaks in this PR.

If this change is really necessary, we could simply convert the query to filters in the chart components themselves. I'd like to avoid converting the internals right now, if possible.

pauldambra replied:
Ah, what a pain... good spot.

I had to stop setting the filters because we default to Trends everywhere

so the sequence was running (something like)

1. get insight X with filters Y
2. init all the things for the linked logics
3. they think they are trends because it is not loaded
4. load it
5. on load of data set the filters on the insight *using the current incorrect defaults*

🤔 I think this might turn out to be a pain....

pauldambra replied:
![concrete case](https://user-images.githubusercontent.com/984817/225565727-ca3e0221-7831-4919-83b0-78f894ad69c4.gif)

If you create a new insight by clicking on the sidebar, you can change tabs fine.

If you create a new insight by clicking on the dropdown and choosing a query based insight you can't change tab

pauldambra commented:
quick addition of showing a text editor when editing an insight

![2023-03-14 21 04 45](https://user-images.githubusercontent.com/984817/225136085-92d7474f-8dd9-451b-aa46-ab77e36241d1.gif)

pauldambra committed:
show text editor for insights

mariusandra reviewed (commented):
Without the `data-exploration-query-tab` flag, things seem a bit broken:

![2023-03-15 16 53 06](https://user-images.githubusercontent.com/53387/225366711-13d43539-0fdc-4953-9ba8-ae73faf6e1f3.gif)

With the flag, we need to update this flow somehow:

![2023-03-15 16 57 28](https://user-images.githubusercontent.com/53387/225367779-cded7f3d-366a-4c61-8934-7ca377b9d7b1.gif)

It's unexpected to end up on a "debug" page after clicking something obvious like "open as a new insight". If nothing else, the title of the tab should be "Data Table", and it should be hidden if not explicitly selected. I'd also hide the code editor by default and make it work like all other insights.

Then,
- I don't know re the question you asked Thomas about spreading values 😅 
- I'd call it not "edit text" but "edit json"... because it's json and our users are adults.
- I'm not fully sold on the icons. Perhaps someone with less than two left hands (talking about myself) wants to suggest some? I think the "open as a new insight" should be a cross between the logo in the sidebar, and a "[open in new window](https://www.google.com/search?q=open+in+new+window+icon)" icon.

pauldambra committed:
Update frontend/src/scenes/insights/InsightPageHeader.tsx

Co-authored-by: Marius Andra <marius.andra@gmail.com>

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
call it json! 🤯

pauldambra committed:
show open editor button for events explorer based on flag combo

pauldambra committed:
just looking at one flag to make life easier

pauldambra committed:
beta flag for json tab

pauldambra commented:
I also have two left hands 🤣 and DALLE isn't helping 

<img width="1300" alt="Screenshot 2023-03-15 at 17 26 59" src="https://user-images.githubusercontent.com/984817/225393399-fef9b7c5-a3b1-4ebf-aa7b-b0fb899953ab.png">

pauldambra commented:
* switched from Query to JSON -> that's really nice 👨‍🍳 👌 
* simplified things
    * if the query flag is on we show
        * query type insights in lists
        * the two query tabs
        * the option to create them
      
## flag on
     
![flag-on-query-tab](https://user-images.githubusercontent.com/984817/225393998-4d87ef9f-718a-4b7a-9bbc-18c07d2e1a3a.gif)

## flag off

![flag-off-query-tab](https://user-images.githubusercontent.com/984817/225394107-0a00122e-921b-4238-ab30-5227f9862f34.gif)

I think the JSON tab needs to stay while this is in flux... but agree a datatable tab should follow this pretty rapidly

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

pauldambra committed:
schema build

pauldambra committed:
one topheading to rule them

mariusandra commented:
I think this is very good! I'd ✅ , but Thomas's point might need a look though :/.

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
fix import

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
no need to connect to data response in insightDataLogic anymore

pauldambra committed:
type check is redundant now

pauldambra committed:
no need for new action/reducer

pauldambra commented:
@mariusandra current plan here is

* get #14787 in to reduce the size of this PR and make the change more obvious
* merge this, it doesn't introduce the bug, just happens to be where we all spotted it

then
* work on the bug in https://github.com/PostHog/posthog/pull/14798 
* check if we can add cypress tests with data exploration flag on and off

pauldambra committed:
query component can ignore null query

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
Merge branch 'master' into feat/replace-open-in-with-insight-page

pauldambra committed:
massage cypress

pauldambra committed:
stickiness couldn't be loaded after lifecycle 🤷

pauldambra committed:
plz cypress plz
